cpp/src/arrow/dataset/file_base.cc

```diff
@@ -265,7 +265,20 @@ Status FileWriter::Write(RecordBatchReader* batches) {
 }
 
 Future<> FileWriter::Finish() {
-  return FinishInternal().Then([this]() { return destination_->CloseAsync(); });
+  return FinishInternal()
+      .Then([this]() {
+        ARROW_ASSIGN_OR_RAISE(bytes_written_, destination_->Tell());
+        return Status::OK();
+      })
+      .Then([this]() { return destination_->CloseAsync(); });
```

Comment:
Nit, but we can probably merge the two `Then` callbacks together?

Reply:
Status implicitly converts to Future, so apart from possibly having to explicitly annotate the return type as Future, you should be able to mix them.